refactor(filetype): extract some functions, use more cache

echasnovski opened the pull request:
There are three parts in this refactor/optimization:
- Separate single "sorted patterns" list into two: "sorted patterns with non-negative priority" and "sorted patterns with negative priority". It allows extracting logic of matching sorted patterns into a separate function. As a result, main `vim.filetype.match()` code is cleaner and pattern matching is easier to manage (hopefully, will be important in future PRs).
- Instead of caching only whether pattern has environment variable (during "match" time), cache other data about it (during "add" time): full pattern (with '^' / '$' at start / end) and whether it contains slash. use them in equivalent manner as before.
- Extract "expand environment variables" in a separate function. I tried simplifying it by simply throwing error if there is no envvar defined (and later call it with `pcall()`), but this seemed to be slower than current approach.
- Convert patterns to be explicit (without implicit anchoring). This proved to be slightly faster (probably due to less backtracking during match) and allows to not compute and cache `fullpat`.

------

Each step also seems to lead to small but consistent performance improvements. I tested (with debug build) for three cases of 'filename' on `main` and latest commit of this PR:

Filename | `master` | This PR |
--------------|-----------|--------------|
init.lua | 0.136ms  | 0.111ms |
no-match | 0.163ms | 0.132ms |
/home/user/.config/hello/world/no-match | 0.240ms | 0.210ms |

------

This is all done in (hopefully) a preparation for #29660.

echasnovski committed:
refactor(filetype): unify matching patterns with pos/neg priority

Problem: due to single list of sorted patterns, their matching inside
  `vim.filetype.match()` was done very similarly but with extra checks
  to stop processing negative priority patterns before extensions.

Solution: create separated sorted lists for patterns with non-negative
  and negative priorities. This allows to process them in a single
  extracted function making the main codeflow a bit nicer and more
  easily expandable.

echasnovski committed:
perf(filetype): cache (more) pattern data during "add" time

gpanders commented on runtime/lua/vim/filetype.lua:
```diff
@@ -4,7 +4,7 @@ local fn = vim.fn
 
 --- @alias vim.filetype.mapfn fun(path:string,bufnr:integer, ...):string?, fun(b:integer)?
---- @alias vim.filetype.maptbl [string|vim.filetype.mapfn, {priority:integer}]
+--- @alias vim.filetype.maptbl { [1]: string|vim.filetype.mapfn, [2]: { priority: number? } }
```
Why change this from the tuple syntax?

echasnovski replied:
Oh, maybe I misunderstanding what LuaLS means by "tuple". Is it indeed the same as table but with known set of indexes? So are these two annotations equivalent?

echasnovski replied:
Yeah, I've updated LuaLS (from 3.7.4 to 3.9.3) and now warnings now.

echasnovski committed:
refactor(filetype): extract expanding env. vars in separate function

echasnovski committed:
refactor(filetype): use Lua patterns without implicit anchoring

echasnovski commented:
@gpanders, I've force pushed several changes that were suggested in #29660. Would you mind re-review, please?

List of changes:
- Caching pattern data is done during "add" time. [Relevant suggestion](https://github.com/neovim/neovim/pull/29660#discussion_r1675514653). Also renamed several local variables to be more explicit.
- All patterns in `pattern` table are now explicit (without implicit anchoring). I planned to suggest it later (after "fast pattern" PR), but here is a [relevant suggestion](https://github.com/neovim/neovim/pull/29660#discussion_r1675510985).

Also rebased on current `master` and updated first comment with more up to date info. Other things are untouched.

lewis6991 commented on runtime/lua/vim/filetype.lua:
```diff
@@ -2218,30 +2221,46 @@ local pattern = {
-    table.insert(sorted, { [k] = { ft, opts } })
+    -- Parse pattern for common data and cache it once
+    pattern_lookup[pat] = pattern_lookup[pat]
+      or { has_env = pat:find('%$%b{}') ~= nil, has_slash = pat:find('/') ~= nil }
```
I'm guessing the point of this is to speed up `vim.filetype.add`?

lewis6991 replied:
On second though, with the PR that adds the grouping, keeping a separate table is better as 99% of the time most of these patterns will never need to be processed.

There still might be merit to keeping this all in one table and lazy populating the fields.

echasnovski replied:
I see.

Is it too much to ask to leave it to some other time (some other contributor)? This whole refactor was to get the necessary pieces for #29660, not make it as perfect as it can be. And at least at the moment I'd prefer to do the reasonable minimum to get the performance gain of #29660 merged.

lewis6991 replied:
I sniped a comment in before yours. We don't need to do this for now, and whatever we do, the parsing should be done lazily.

lewis6991 commented on runtime/lua/vim/filetype.lua:
```diff
@@ -2218,30 +2221,46 @@ local pattern = {
-  return sorted
+
+  table.sort(pos, compare_by_priority)
+  table.sort(neg, compare_by_priority)
```
Light suggestion: we could just sort the built-in table once, then any additions with `vim.filetype.add` can place the new filetypes at the correct position rather than invoking a full sort.

echasnovski replied:
That is a reasonable thing to do. I'd leave it for some other time.

clason replied:
At least add a todonote for it.

echasnovski replied:
It's already approved, so I don't want to force push. Is it OK to leave as is? It would still pop up if/when `vim.filetype.add()` optimization is considered to be done.

clason replied:
I'd still prefer it (I don't know where the force-push concern comes from.)

echasnovski replied:
OK. I'll add it shortly.

echasnovski replied:
Added a TODO comment.

lewis6991 commented:
Ready to merge?

echasnovski commented:
> Ready to merge?

From my side - yes.
